Karapace schema registry (#1342)
aiven/devportal: docs/products/kafka/karapace/concepts/schema-registry-authorization.rst

Common use cases
----------------
Some of the common use cases for Karapace schema registry authorization include: 

* **Schemas as API contract among teams:** Allow teams to leverage schemas as API contracts with other teams by providing read-only access to schemas in Schema Registry.
* **Kafka as a communication broker with third parties:** Allow conditional read-only access to schemas by third parties to establish Kafka topics as communication channels.
* **Schema registration automated with CI/CD:** Achieve automation of schema registration using Continuous Integration(CI) tools in such a way that add, update and delete operations on schemas is limited to the CI Tools. At the same time, the different applications are allowed read-only access to the schemas as needed.
* **Segregated consumers of Schema Registry and REST Proxy:** Limits the attack surface by segregating and limiting access to only what is required for the consumers of the Schema Registry and REST Proxy.

.. _karapace_schema_registry_acls:

> ACLs definition

Reviewer: Should this be on a separate `concept` file?

---------------
An ACL for Karapace schema registry authorization consists of zero or more entries that specify a **username**, an **operation**, and a **resource**. 

* **Username** is the name of a service user in the Aiven for Apache Kafka® service.
* **Operations** are:
  
  * ``schema_registry_read``

>   * ``schema_registry_write`` (always includes schema_registry_read)

Reviewer: ```suggestion
  * ``schema_registry_write`` (always includes ``schema_registry_read``)
```

  
   * ``Subject:subject_name``: the entry controls access to subjects in the schema registry.

.. Tip::

  The ``name`` specified in the ACL entry resource and username can use wildcards:
      
  * ``*`` matching any characters
  * ``?`` matching a single character

When a user requests access to a resource, the schema registry checks to see if any entry in the ACL matches the requesting user and the requested resource, to decide whether to grant or deny read/write access to the requested resource. The order of entries in the ACL does not affect the access control decision.

The schema registry responds with HTTP status code 401 Unauthorized message if no ACL entries grant access.

.. Note:: 
  Enabling or disabling Karapace schema registry authorization, and managing the ACLs, is done using the Aiven CLI (requires version 2.16 or later). For more information, see :doc:`Enable Karapace schema registry authorization <../howto/enable-schema-registry-authorization>`. 
 

Securing endpoints with ACL
---------------------------

To correctly set up access control for the different endpoints, you need to have the following in the ACL:

> * Endpoints that provide read-only operations require schema_registry_read permission in the ACL for the specific subject.  In the case of endpoints that return data related to multiple subjects, the response is filtered to include data against only those subjects the user has been granted read permission. 

Reviewer: ```suggestion
* Endpoints that provide read-only operations require ``schema_registry_read`` permission in the ACL for the specific subject.  In the case of endpoints that return data related to multiple subjects, the response is filtered to include data against only those subjects the user has been granted read permission. 
```

Karapace schema registry authorization
======================================
The schema registry authorization feature when enabled in :doc:`Karapace schema registry <../getting-started>` allows you to authenticate the user, and control read or write access to the individual resources available in the Schema Registry. 
Authorization in Karapace is achieved by using :doc:`Access Control Lists (ACLs) <../concepts/acl-definition>`. ACLs provide a way to achieve fine-grained access control for the resources in Karapace.

.. Note::

>   Karapace schema registry authorization is enabled on all Aiven for Apache Kafka® services. The exception is older services created before mid-2022, where the feature needs to be :doc:`enabled <../getting-started>`.

Reviewer: Link to the enable doc?
```suggestion
  Karapace schema registry authorization is enabled on all Aiven for Apache Kafka® services. The exception is older services created before mid-2022, where the feature needs to be :doc:`enabled <../howto/enable-karapace>`.
```

Common use cases
----------------
Some of the common use cases for Karapace schema registry authorization include: 

* **Schemas as API contract among teams:** Allow teams to leverage schemas as API contracts with other teams by providing read-only access to schemas in Schema Registry.
* **Kafka as a communication broker with third parties:** Allow conditional read-only access to schemas by third parties to establish Kafka topics as communication channels.
* **Schema registration automated with CI/CD:** Achieve automation of schema registration using Continuous Integration(CI) tools in such a way that add, update and delete operations on schemas is limited to the CI Tools. At the same time, the different applications are allowed read-only access to the schemas as needed.
* **Segregated consumers of Schema Registry and REST Proxy:** Limits the attack surface by segregating and limiting access to only what is required for the consumers of the Schema Registry and REST Proxy.

> .. _karapace_schema_registry_acls:

Reviewer 2: here is how you add a label @harshini-rangaswamy
